### PR TITLE
report: fix element screenshot position, lifecycle, styles

### PR DESCRIPTION
**Pull request by paulirish**
this is an alternative to #11843  and #11844

(and here's the diff against #11843, which may be useful: https://github.com/GoogleChrome/lighthouse/compare/fps-tweak-viewport...fps-tweak-partytime )

okay a few things..

### use fixed positioning to keep the overlay on top of the report content

if our overlay element was placed higher up in the devtools DOM, `pos:abs` would work. (That's what it used there for [their equivalent "glassPane" component](https://source.chromium.org/chromium/chromium/src/+/master:third_party/devtools-frontend/src/front_end/ui/glassPane.css;l=7-16;bpv=1;bpt=0)).  

i don't see a good way to use `pos:sticky` without a nasty hack like `margin-bottom: -100vh;` to fix it [pushing down the report content](https://github.com/GoogleChrome/lighthouse/pull/11843/files#r544688266). 

`fixed` works well. DOM-wise the position of the overlay doesn't matter since it rolls up to its containing block. It can remain a child of `.lh-report` so there's no diff, but I slightly prefer it's a child of `lh-root` so it has more visibility.

#### overlay positioning aside

There are options but it gets interesting when considering our report within another page/webapp. (eg DevTools). 

1. does overlay go on top of topbar? I think yes, cuz otherwise you can toggle the topbar menu and it displays _under_ the overlay
2. does overlay go on top of devtools toolbar?  I wasn't sure but I checked the DevTools glasspane and it goes on top. Here's an example modal & glasspane in Network panel: rightclick table headers > response headers > manage header columns. (Also their glasspane is transparent, so I've done that too.) Anyway, this all convinced me that the overlay should sit on top of everything. 

### size the lightbox screenshot relative to the overlay

with the overlay the size we want, we can use its dimensions as input for the `maxLightboxSize` which goes into the amazing sizing algorithm. that thing is too sweet and handles landscape/portrait viewports so so well.

this means we add the overlay to the dom first and then measure it, but i think that's OK!

### small style tweaks on lightbox overlay.

it had 20px margin that it wanted for the thumbnail but was being applied in the lightbox too:
image

also i added a 1px `outline` around it, just for visual contrast. (yay for outline not affecting dimensions)

and I lessened the overlay's gray bgcolor a tad. more opacity ~= less dark. and in devtools, it's transparent to match the glasspane.

(obv these are not critical and i can defer till later if desired)

### fix lightbox click lifecycle

basically the same thing that #11845 was getting at.  The core problem is that we use the `.lh-element-screenshot` class for both thumbnails and lightbox images. 

I kept @connorjclark's "moved the listener to the rootEl, instead of the document". 👍 

And after starting with adding an extra `if elem.closest(am-i-in-the-lightbox) return;` I realized [the existing selector could be tweaked](880fee527e5620b4818a91bab8abfb2170c5ed37).

### CODE REVIEW

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 1]
```diff
@@ -139,24 +139,30 @@ class ElementScreenshotRenderer {
    * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
    */
   static installOverlayFeature(dom, templateContext, fullPageScreenshot) {
-    const reportEl = dom.find('.lh-report', dom.document());
-    const screenshotOverlayClass = 'lh-feature-screenshot-overlay';
-    if (reportEl.classList.contains(screenshotOverlayClass)) return;
-    reportEl.classList.add(screenshotOverlayClass);
+    const topbarEl = dom.find('.lh-topbar', dom.document());
+    const containerEl = topbarEl.parentElement;
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js]: want to add the explanation you had over VC for why it conceptually belongs next to the topbar? :)

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 2]
```diff
@@ -139,24 +139,30 @@ class ElementScreenshotRenderer {
...
+    // Don't install the feature more than once.
+    if (containerEl.classList.contains(screenshotOverlayClass)) return;
+    containerEl.classList.add(screenshotOverlayClass);
 
-    dom.document().addEventListener('click', e => {
+    containerEl.addEventListener('click', e => {
       const target = /** @type {?HTMLElement} */ (e.target);
       if (!target) return;
-      const el = /** @type {?HTMLElement} */ (target.closest('.lh-element-screenshot'));
+      const el = /** @type {?HTMLElement} */ (target.closest('.lh-node > .lh-element-screenshot'));
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:156]:
```suggestion
      // Only activate the overlay for clicks on the screenshot *preview* of an element, not the full-size too.
      const el = /** @type {?HTMLElement} */ (target.closest('.lh-node > .lh-element-screenshot'));
```

right?

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 3]
```diff
@@ -139,24 +139,30 @@ class ElementScreenshotRenderer {
...
-      width: dom.document().documentElement.clientWidth,
-      height: dom.document().documentElement.clientHeight * 0.75,
-    };
+    const screenshotOverlayClass = 'lh-feature-screenshot-overlay';
+    // Don't install the feature more than once.
+    if (containerEl.classList.contains(screenshotOverlayClass)) return;
+    containerEl.classList.add(screenshotOverlayClass);
 
-    dom.document().addEventListener('click', e => {
+    containerEl.addEventListener('click', e => {
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js]:
```suggestion
    // Add a single listener to the top container to handle a click on any node.
    containerEl.addEventListener('click', e => {
```

some optional explicitness :)

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 4]
```diff
@@ -139,24 +139,30 @@ class ElementScreenshotRenderer {
...
       const target = /** @type {?HTMLElement} */ (e.target);
       if (!target) return;
-      const el = /** @type {?HTMLElement} */ (target.closest('.lh-element-screenshot'));
+      const el = /** @type {?HTMLElement} */ (target.closest('.lh-node > .lh-element-screenshot'));
       if (!el) return;
 
-      const overlay = dom.createElement('div');
-      overlay.classList.add('lh-element-screenshot__overlay');
+      const overlay = dom.createElement('div', 'lh-element-screenshot__overlay');
+      containerEl.insertBefore(overlay, topbarEl.nextElementSibling);
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js]: does the `insertBefore` still have significance if we're back to position fixed or can it just be appendChild?

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 5]
```diff
@@ -172,14 +178,13 @@ class ElementScreenshotRenderer {
         elementRectSC,
         maxLightboxSize
       );
-      if (!screenshotElement) return;
 
+      // This would be unexpected here. The null return is used to not create a thumbnail in details-renderer
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:184]: I'm not sure I understand the null *return* bit. is this saying `screenshotElement` is null only when there isn't a thumbnail so they shouldn't have been able to click on anything to get here?

**Review by patrickhulce — COMMENTED** [review]
I like this approach 👍 

@connorjclark how do you feel about this alternative?

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js (hunk 1 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/element-screenshot-renderer.js]: since your comment I changed this. and at the top i added this explanation

> DOM-wise the position of the overlay doesn't matter since it rolls up to its containing block. It can remain a child of .lh-report so there's no diff, but I slightly prefer it's a child of lh-root so it has more visibility.

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js (hunk 5 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/element-screenshot-renderer.js:185]: yes that's accurate.

Comment (patrickhulce) [reply lighthouse-core/report/html/renderer/element-screenshot-renderer.js:186]:
```suggestion
      // This would be unexpected here. 
      // When `screenshotElement` is `null`, there is also no thumbnail element for the user to have clicked to make it this far. 
```

how do you feel about this? :D

**Review by Technotips786 — COMMENTED** [review]
I can not understand it.

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 6]
```diff
@@ -139,24 +139,34 @@ class ElementScreenshotRenderer {
...
-
-    const maxLightboxSize = {
-      width: dom.document().documentElement.clientWidth,
-      height: dom.document().documentElement.clientHeight * 0.75,
-    };
+    const rootEl = dom.find('.lh-root', dom.document());
+    if (!rootEl) {
+      return console.warn('No lh-root. Overlay install failed.'); // eslint-disable-line no-console
+    }
+
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:146]:
```suggestion
```

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 7]
```diff
@@ -139,24 +139,34 @@ class ElementScreenshotRenderer {
...
-    if (reportEl.classList.contains(screenshotOverlayClass)) return;
-    reportEl.classList.add(screenshotOverlayClass);
-
-    const maxLightboxSize = {
-      width: dom.document().documentElement.clientWidth,
-      height: dom.document().documentElement.clientHeight * 0.75,
-    };
+    const rootEl = dom.find('.lh-root', dom.document());
+    if (!rootEl) {
+      return console.warn('No lh-root. Overlay install failed.'); // eslint-disable-line no-console
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:144]: nit: these `return <void>` calls give me the heebee-jeebies. reminds me of the days of express apps and callback hell. consider `return` on a separate line?

it is also a pattern that I don't think we do anywhere else.

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 8]
```diff
@@ -139,24 +139,34 @@ class ElementScreenshotRenderer {
...
+    }
+
+
+    const screenshotOverlayClass = 'lh-screenshot-overlay--enabled';
+    // Don't install the feature more than once.
+    if (rootEl.classList.contains(screenshotOverlayClass)) return;
+    rootEl.classList.add(screenshotOverlayClass);
 
-    dom.document().addEventListener('click', e => {
+    // Add a single listener to the top container to handle all clicks within (event delegation)
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:153]:
```suggestion
    // Add a single listener to the root element to handle all clicks within (event delegation).
```

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 9]
```diff
@@ -139,24 +139,34 @@ class ElementScreenshotRenderer {
...
+    rootEl.addEventListener('click', e => {
       const target = /** @type {?HTMLElement} */ (e.target);
       if (!target) return;
-      const el = /** @type {?HTMLElement} */ (target.closest('.lh-element-screenshot'));
+      // Only activate the overlay for clicks on the screenshot *preview* of an element, not the full-size too.
+      const el = /** @type {?HTMLElement} */ (target.closest('.lh-node > .lh-element-screenshot'));
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:158]:
```suggestion
      const thumbnailEl = /** @type {?HTMLElement} */ (target.closest('.lh-node > .lh-element-screenshot'));
```

maybe says same thing but without a comment?

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 10]
```diff
@@ -139,24 +139,34 @@ class ElementScreenshotRenderer {
...
       if (!el) return;
 
-      const overlay = dom.createElement('div');
-      overlay.classList.add('lh-element-screenshot__overlay');
+      const overlay = dom.createElement('div', 'lh-element-screenshot__overlay');
+      rootEl.append(overlay);
+
+      // The newly-added overlay has the bounding rect we need
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:164]:
```suggestion
      // The newly-added overlay has the dimensions we need.
```

File: lighthouse-core/report/html/report-styles.css [hunk 11]
```diff
@@ -104,6 +104,7 @@
   --image-preview-size: 48px;
   --metric-toggle-lines-fill: #7F7F7F;
   --metrics-toggle-background-color: var(--color-gray-200);
+  --modal-overlay-background: rgba(0, 0, 0, 0.3);
```
Comment (connorjclark) [inline lighthouse-core/report/html/report-styles.css:107]: `--screenshot-overlay-background`? or are you anticipating future modals?

**Review by connorjclark — COMMENTED** [review]
QAing now. flushing comments.

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 12]
```diff
@@ -139,24 +139,34 @@ class ElementScreenshotRenderer {
...
+
+
+    const screenshotOverlayClass = 'lh-screenshot-overlay--enabled';
+    // Don't install the feature more than once.
+    if (rootEl.classList.contains(screenshotOverlayClass)) return;
+    rootEl.classList.add(screenshotOverlayClass);
 
-    dom.document().addEventListener('click', e => {
+    // Add a single listener to the top container to handle all clicks within (event delegation)
+    rootEl.addEventListener('click', e => {
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:154]: (not now, but for later)

should we listen to resize so we can re-make an open overlay? things get messy when you dock/undock, but also impacts just resizing the window of a standalone report

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js (hunk 9 above)
Comment (patrickhulce) [reply lighthouse-core/report/html/renderer/element-screenshot-renderer.js:158]: I'm to blame for the comment :) WFM 👍

**Comment by connorjclark** [comment]
image
is this expected (not being vertically centered)?

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js (hunk 9 above)
Comment (paulirish) [reply lighthouse-core/report/html/renderer/element-screenshot-renderer.js:158]: one way this'll clean up is if we use `.lh-element-screenshot--thumbnail`  and `.lh-element-screenshot--lightbox` everywhere. i can do this in a followup

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 13]
```diff
@@ -172,14 +182,15 @@ class ElementScreenshotRenderer {
         elementRectSC,
         maxLightboxSize
       );
-      if (!screenshotElement) return;
 
-      overlay.appendChild(screenshotElement);
-      overlay.addEventListener('click', () => {
+      // This would be unexpected here.
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:186]:
```suggestion
      // This would be unexpected here.
```

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 14]
```diff
@@ -172,14 +182,15 @@ class ElementScreenshotRenderer {
         elementRectSC,
         maxLightboxSize
       );
-      if (!screenshotElement) return;
 
-      overlay.appendChild(screenshotElement);
-      overlay.addEventListener('click', () => {
+      // This would be unexpected here. 
+      // When `screenshotElement` is `null`, there is also no thumbnail element for the user to have clicked to make it this far.
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:187]:
```suggestion
      // When `screenshotElement` is `null`, there is also no thumbnail element for the user to have clicked to make it this far.
```

**Review by paulirish — COMMENTED** [review]
> is this expected (not being vertically centered)?


it _is_ vertically centered. 😄 

(random image edits to prove the same amount of pixels on bottom as top:

image

.. vertically centered assuming the "overlay positioning aside" bit from the description"

**Review by connorjclark — APPROVED** [review]

**Comment by paulirish** [comment]
@Technotips786 i'm looking for something halfway between the prodigy and orbital. any suggestions?
